### DISCUSSION TITLE
uint16_t getTotalFolderCount() results in 0

### DISCUSSION THREAD

**Discussion by RobertAngermayr** [Makuna/DFMiniMp3#133] [Q&A]
I have a MH2024K 24SS player (labeld as MP3-TF-16P V3.0)

the function getTotalFolderCount always results in 0 folders.
Also i cannot find the Mp3_Commands_GetTotalFolderCount = 0x4f in any documentation.
All the documents i found stop at 0x4D as last command.

All the other commands seem to work, as far i've tested it.

**Comment by Makuna** [comment]
If you turn on the [debug feature](https://github.com/Makuna/DFMiniMp3/wiki/Debug-Features) it will give us more details on the communications going on for that API.
Also, if you look in the examples for this library, in the extras folder you will find a test sketch Mp3_Ack_Api_Test.ino that runs through every API call I fully support.

The Flyron FN-M10P documentation has both 0x4e and 0x4f in it.
https://www.electronics123.net/amazon/datasheet/FN-M10P%20Embedded%20MP3%20Audio%20Module%20Datasheet.pdf
This was one of the first data sheets available.  
The MH2024 came along late in the MP3 modules, and it has several variants (all labeled the same) that certain other calls don't work the same between them.  Unfortunately, this makes it hard for customers to really know what they have and what I can support or not support.

There are a few others with varying degrees of support.
Mp3_Commands_GetPlaySources = 0x3f : so bad, I depreciated the call for it as it overlaps a notification also.
0x10:  A volume command that did nothing.  To this day I still don't know what it was meant to be used for and some specs don't show it.
Mp3_Commands_LoopGlobalTrack = 0x08 : same as below, more often the supported action, which means you can query but not set playback mode?
Mp3_Commands_SetPlaybackMode = 0x08 : save as above, a few early chips used this action
Mp3_Commands_GetStatus = 0x42 : was horribly documented (some had it, some had it wrong), but the latest version of this library should be handling it correctly.

**Comment by RobertAngermayr** [comment]
Here is the reply with debug on.

what im wondering is that the OUT command is sent three times.
And when i ask for the number of files in a folder its asked only one times, than a result appears.

initializing...
OUT  7e ff 06 0c 01 00 00 fe ee ef
IN  7e ff 06 41 00 00 00 fe ba ef
IN  7e ff 06 3f 00 00 02 fe ba ef
SD Card, online
starting...
-------------------------------
OUT  7e ff 06 4e 00 00 02 fe ab ef
IN  7e ff 06 4e 00 00 32 fe 7b ef
number of files in folder 2 is 50
-------------------------------
OUT  7e ff 06 4f 00 00 00 fe ac ef
OUT  7e ff 06 4f 00 00 00 fe ac ef
OUT  7e ff 06 4f 00 00 00 fe ac ef
number of folders is 0
-------------------------------
OUT  7e ff 06 4e 00 00 02 fe ab ef
IN  7e ff 06 4e 00 00 32 fe 7b ef
number of files in folder 2 is 50
-------------------------------

the init code:
`
#define delayBetweenCommands 100

void setup() {

  Serial.begin(115200);
  delay(5000);
  Serial.println("initializing...");
  Serial1.begin(9600);
  delay(1000);
  dfmp3.begin(9600);
  waitMilliseconds(delayBetweenCommands);

  dfmp3.reset(); 
  waitMilliseconds(delayBetweenCommands);

  Serial.println("starting...");

  uint16_t fon;    

  Serial.println ("-------------------------------");
  fon = dfmp3.getFolderTrackCount(2);
  delay(100);
  waitMilliseconds(delayBetweenCommands);
  Serial.print("number of files in folder 2 is "); 
  Serial.println ( fon );

  Serial.println ("-------------------------------");
  fon = dfmp3.getTotalFolderCount();
  delay(100);
  waitMilliseconds(delayBetweenCommands);
  Serial.print("number of folders is "); 
  Serial.println ( fon );

  Serial.println ("-------------------------------");
  fon = dfmp3.getFolderTrackCount(2);
  delay(100);
  waitMilliseconds(delayBetweenCommands);
  Serial.print("number of files in folder 2 is "); 
  Serial.println ( fon );

  Serial.println ("-------------------------------");  
}

void waitMilliseconds(uint16_t msWait)
{
  uint32_t start = millis();
  
  while ((millis() - start) < msWait)
  {
    // if you have loops with delays, its important to 
    // call dfmp3.loop() periodically so it allows for notifications 
    // to be handled without interrupts
    dfmp3.loop(); 
    delay(1);
  }
}
`

**Comment by Makuna** [comment]
It should trigger an error callback with RX Timeout.  But it does seem to just ignore the request.

**Comment by RobertAngermayr** [comment]
That's another question, with the dfplayer library I receive the timeout, I enlarged the timeout to one minute but still timeout.
With your library I always receive error -1 no difference between the possible failures.

**Reply by Makuna** [reply]
If nothing comes back from the chip when we expect it, you get a timeout error.

[Wiki entry for OnError()](https://github.com/Makuna/DFMiniMp3/wiki/Notification-Method#static-void-onerrordfmp3-mp3-uint16_t-errorcode)
-1?  The callback for errors is uint16_t, so it should be `DfMp3_Error_RxTimeout` which is 0x81, which if cast to an int8_t would be -1; but it's not a signed value.

```
static void OnError(DfMp3& mp3, uint16_t errorCode)
```
